lib/DataFileUtil/DataFileUtilImpl.py

```diff
@@ -1082,6 +1358,98 @@ def versions(self, ctx):
+           staging_file_subdir_path is subdir_1/subdir_2/file_name) ->
+           structure: parameter "staging_file_subdir_path" of String
+        :returns: instance of type "DownloadStagingFileOutput" (Results from
+           the download_staging_file function. copy_file_path: copied file
+           scratch area path) -> structure: parameter "copy_file_path" of
+           String
+        """
+        # ctx is the context object
+        # return variables are: results
+        #BEGIN download_staging_file
+        if not params.get('staging_file_subdir_path'):
+          error_msg = "missing 'staging_file_subdir_path' parameter"
+          raise ValueError(error_msg)
+        
+        staging_file_subdir_path = params.get('staging_file_subdir_path')
+        staging_file_name = os.path.basename(staging_file_subdir_path)
+        staging_file_path = self._get_staging_file_path(
+                                    ctx['user_id'], staging_file_subdir_path)
+
+        self.log('Start downloading staging file: %s' % staging_file_path)
+        shutil.copy2(staging_file_path, self.tmp)
+        copy_file_path = os.path.join(self.tmp, staging_file_name)
+        self.log('Copied staging file from %s to %s' %
+                 (staging_file_path, copy_file_path))
+
+        copy_file_path = self._unpack(copy_file_path, True)
+
+        results = {'copy_file_path': copy_file_path}
+        #END download_staging_file
+
+        # At some point might do deeper type checking...
+        if not isinstance(results, dict):
+            raise ValueError('Method download_staging_file return value ' +
+                             'results is not type dict as required.')
+        # return the results
+        return [results]
+
+    def download_web_file(self, ctx, params):
+        """
+        Download a web file to scratch area
+        :param params: instance of type "DownloadWebFileParams" (Input
+           parameters for the "download_web_file" function. Required
+           parameters: file_url: file URL download_type: one of ['Direct
+           Download', 'FTP', 'DropBox', 'Google Drive']) -> structure:
+           parameter "file_url" of String, parameter "download_type" of String
+        :returns: instance of type "DownloadWebFileOutput" (Results from the
+           download_web_file function. copy_file_path: copied file scratch
+           area path) -> structure: parameter "copy_file_path" of String
+        """
+        # ctx is the context object
+        # return variables are: results
+        #BEGIN download_web_file
+        
+        # check for required parameters
+        for p in ['file_url', 'download_type']:
+          if p not in params:
+            raise ValueError("missing '{}' parameter".format(p)) 
+
+        file_url = params.get('file_url')
+        download_type = params.get('download_type')
+
+        if download_type != 'FTP':
+          try:
+            response = requests.get(file_url, stream=True)
```

Comment:
This check occurs here but also when the file name is determined from the url or content-disp header and when the actual file is downloaded. This seems redundant. Also, in some cases the url is modified prior to getting the filename - couldn't this cause problems by trying to access the unmodified url?